python/ray/tune/syncer.py

```diff
@@ -28,198 +28,256 @@
+        self._validate_sync_string(sync_up_template)
+        self._validate_sync_string(sync_down_template)
+        self.sync_up_template = sync_up_template
+        self.sync_down_template = sync_down_template
+        self.logfile = None
+        self.sync_process = None
 
-class CommandSyncer(BaseSyncer):
-    def __init__(self, local_dir, remote_dir, sync_template):
-        """Syncs between two directories with the given command.
+    def set_logdir(self, logdir):
+        """Sets the directory to log sync execution output in.
 
-        Arguments:
-            local_dir (str): Directory to sync.
-            remote_dir (str): Remote directory to sync with.
-            sync_template (str): A string template
-                for syncer to run and needs to include replacement fields
-                '{source}' and '{target}'. Returned when using
-                `CommandSyncer.sync_template`, which can be overridden
-                by subclass.
+        Args:
+            logdir: Log directory.
         """
-        super(CommandSyncer, self).__init__(local_dir, remote_dir)
-        if not isinstance(sync_template, str):
-            raise ValueError("{} is not a string.".format(sync_template))
-        validate_sync_string(sync_template)
-        self._sync_template = sync_template
         self.logfile = tempfile.NamedTemporaryFile(
-            prefix="log_sync",
-            dir=self._local_dir,
-            suffix=".log",
-            delete=False)
+            prefix="log_sync", dir=logdir, suffix=".log", delete=False)
 
-        self.sync_process = None
+    def sync_up(self, source, target):
+        return self.execute(self.sync_up_template, source, target)
+
+    def sync_down(self, source, target):
+        return self.execute(self.sync_down_template, source, target)
 
-    def sync_function(self, source, target):
-        self.last_sync_time = time.time()
+    def execute(self, sync_template, source, target):
+        """Executes sync_template on source and target."""
         if self.sync_process:
             self.sync_process.poll()
             if self.sync_process.returncode is None:
                 logger.warning("Last sync is still in progress, skipping.")
-                return
-        final_cmd = self._sync_template.format(
+                return False
+        final_cmd = sync_template.format(
             source=quote(source), target=quote(target))
         logger.debug("Running sync: {}".format(final_cmd))
         self.sync_process = subprocess.Popen(
-            final_cmd, shell=True, stdout=self.logfile)
+            final_cmd, shell=True, stderr=subprocess.PIPE, stdout=self.logfile)
         return True
 
+    def wait(self):
+        if self.sync_process:
+            _, error_msg = self.sync_process.communicate()
+            code = self.sync_process.returncode
+            self.sync_process = None
+            if code != 0:
+                raise TuneError("Sync error ({}): {}".format(code, error_msg))
```

Comment:
this is returned as byte literal right? I remember this printed oddly.

Reply:
yeah actually not sure why, most of the time it prints fine but i saw a big byte dump once or twice too